# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/render_task.rs
# Commit: e38dce4a5b06
# Full Hash: e38dce4a5b06869dd4a4d753f41f7594a8cddfb6
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2021-04-22 21:31:57
# Regressor Bug: 1702228
# File Overlap Count: 2
# Description:
#   Bug 1702228 - Cache linear gradients by default. r=gw
#   
#   This patch breaks linear gradients into two primitives:
#    - LinearGradient is always rendered via the brush shader. It is only used with SWGL.
#    - CachedLinearGradient is always rendered via a cached render task and the brush image shader. Its implementation is very similar to conic and radial gradients, with the addition of a fast-path for axis aligned gradients with two stops.
# ==============================================================================

diff -r b3f8befa510f -r e38dce4a5b06 gfx/wr/webrender/src/render_task.rs
--- a/gfx/wr/webrender/src/render_task.rs	Thu Apr 22 10:34:54 2021 +0000
+++ b/gfx/wr/webrender/src/render_task.rs	Thu Apr 22 10:34:54 2021 +0000
@@ -16,8 +16,8 @@
 use crate::picture::{ResolvedSurfaceTexture, SurfaceInfo};
 use crate::prim_store::{ClipData, PictureIndex};
 use crate::prim_store::gradient::{
-    GRADIENT_FP_STOPS, GradientStopKey, FastLinearGradientTask, RadialGradientTask,
-    ConicGradientTask,
+    FastLinearGradientTask, RadialGradientTask,
+    ConicGradientTask, LinearGradientTask,
 };
 #[cfg(feature = "debugger")]
 use crate::print_tree::{PrintTreePrinter};
@@ -307,6 +307,7 @@
     Border(BorderTask),
     LineDecoration(LineDecorationTask),
     FastLinearGradient(FastLinearGradientTask),
+    LinearGradient(LinearGradientTask),
     RadialGradient(RadialGradientTask),
     ConicGradient(ConicGradientTask),
     SvgFilter(SvgFilterTask),
@@ -338,6 +339,7 @@
             RenderTaskKind::Border(..) => "Border",
             RenderTaskKind::LineDecoration(..) => "LineDecoration",
             RenderTaskKind::FastLinearGradient(..) => "FastLinearGradient",
+            RenderTaskKind::LinearGradient(..) => "LinearGradient",
             RenderTaskKind::RadialGradient(..) => "RadialGradient",
             RenderTaskKind::ConicGradient(..) => "ConicGradient",
             RenderTaskKind::SvgFilter(..) => "SvgFilter",
@@ -353,6 +355,7 @@
             RenderTaskKind::Readback(..) |
             RenderTaskKind::Border(..) |
             RenderTaskKind::FastLinearGradient(..) |
+            RenderTaskKind::LinearGradient(..) |
             RenderTaskKind::RadialGradient(..) |
             RenderTaskKind::ConicGradient(..) |
             RenderTaskKind::Picture(..) |
@@ -412,20 +415,6 @@
         })
     }
 
-    pub fn new_gradient(
-        stops: [GradientStopKey; GRADIENT_FP_STOPS],
-        orientation: LineOrientation,
-        start_point: f32,
-        end_point: f32,
-    ) -> Self {
-        RenderTaskKind::FastLinearGradient(FastLinearGradientTask {
-            stops,
-            orientation,
-            start_point,
-            end_point,
-        })
-    }
-
     pub fn new_readback(
         readback_origin: Option<DevicePoint>,
     ) -> Self {
@@ -634,6 +623,7 @@
             RenderTaskKind::Border(..) |
             RenderTaskKind::LineDecoration(..) |
             RenderTaskKind::FastLinearGradient(..) |
+            RenderTaskKind::LinearGradient(..) |
             RenderTaskKind::RadialGradient(..) |
             RenderTaskKind::ConicGradient(..) |
             RenderTaskKind::Blit(..) => {
@@ -1460,6 +1450,9 @@
             RenderTaskKind::FastLinearGradient(..) => {
                 pt.new_level("FastLinearGradient".to_owned());
             }
+            RenderTaskKind::LinearGradient(..) => {
+                pt.new_level("LinearGradient".to_owned());
+            }
             RenderTaskKind::RadialGradient(..) => {
                 pt.new_level("RadialGradient".to_owned());
             }